File: Difficulty/Medium/[89] Gray Code/solution.py
class Solution:
    def grayCode(self, n: int) -> List[int]:
        num = 1 << n
        res = [0] * num
        for i in range(1, num):
            if i & 1:
                res[i] = res[i - 1] ^ 1 if res[i - 1] & 1 else res[i - 1] | 1
            else:
                j = self.getPosition(i)
                res[i] = res[i - 1] ^ (1 << (j - 1)) if res[i - 1] & (1 << (j - 1)) else res[i - 1] | (1 << (j - 1))
        return res
    
    def getPosition(self, n):
        i = 2
        while not (n >> 1) & 1:
            n >>= 1 
            i += 1
        return i

File: Difficulty/Medium/[89] Gray Code/solution.cpp
/*
 * @lc app=leetcode id=89 lang=cpp
 *
 * [89] Gray Code
 */
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> grayCode(int n) {
        int num = 1 << n;
        vector<int> res(num, 0);
        for (int i = 1; i < num; ++i) {
            if (i & 1)
                res[i] = res[i - 1] & 1 ? res[i - 1] ^ 1 : res[i - 1] | 1;
            else
            {
                int j = getPosition(i);
                res[i] = res[i - 1] & (1 << (j - 1)) ? res[i - 1] ^ (1 << (j - 1)) : res[i - 1] | (1 << (j - 1));
            }
        }
        return res;
    }
    int getPosition(int n){
        int i = 2;
        while (! ((n = n >> 1) & 1)) i++;
        return i;
    }
};

